tfhe/src/integer/server_key/radix_parallel/scalar_div_mod.rs

```diff
@@ -0,0 +1,481 @@
+
+        // The formula is m_low == 2**(N+l) / d
+        // however N + l may be == to 128, which means doing
+        // 2**(N + l) would not be valid, so the formula has been written in a
+        // way that does not overlflow.
+        //
+        // By writing m_low = (m_low - 2**N) + (2**N)
+        let mut m_low = (two_pow_n * ((1 << l) - d)) / d;
+        m_low += two_pow_n;
+
+        // The formula is (2**(N+l) + 2**(N + l - precision)) / d
+        // again 2**(N+l) could overflow
+        let mut m_high = ((two_pow_n * ((1 << l) - d)) + (1 << (N + l - precision))) / d;
+        m_high += two_pow_n;
+
+        assert!(m_low < m_high);
+
+        loop {
+            let m_low_i = m_low >> 1;
+            let m_high_i = m_high >> 1;
+
+            if m_low_i >= m_high_i || shift_post == 0 {
+                break;
+            }
+
+            m_low = m_low_i;
+            m_high = m_high_i;
+            shift_post -= 1;
+        }
+
+        Self {
+            multiplier: m_high,
+            shift_post,
+        }
+    }
+}
+
+impl ServerKey {
+    /// computes lhs * rhs on the full precision (num_block * 2) and returns
+    /// the most significant num blocks.
+    ///
+    /// In other words, it is like doing
+    /// fn mulhi (a: u32, b: u32) -> u32 {
+    ///    (a as 64 * b as u64) >> 32) as u32
+    /// }
+    fn scalar_mul_high(&self, lhs: &RadixCiphertext, rhs: u128) -> RadixCiphertext {
+        let mut result = lhs.clone();
+        self.extend_radix_with_trivial_zero_blocks_msb_assign(&mut result, lhs.blocks.len());
+        self.scalar_mul_assign_parallelized(&mut result, rhs);
+        self.trim_radix_blocks_lsb_assign(&mut result, lhs.blocks.len());
+        result
+    }
+
+    pub fn unchecked_scalar_div_parallelized<T>(
+        &self,
+        numerator: &RadixCiphertext,
+        divisor: T,
+    ) -> RadixCiphertext
+    where
+        u64: From<T>,
+        T: Numeric,
+    {
+        let divisor_bits = T::BITS as u32;
+        let numerator_bits = self.key.message_modulus.0.ilog2() * numerator.blocks.len() as u32;
+        assert!(
+            divisor_bits <= numerator_bits,
+            "The divisor has more bits than the numerator\
+                (numerator: {numerator_bits} bits, divisor: {divisor_bits} bits)"
+        );
+
+        let divisor = u64::from(divisor);
+        // Rust has a check on all devision, so shall also have one
+        assert_ne!(divisor, 0, "Cannot divide by zero");
+
+        if divisor.is_power_of_two() {
+            // Even in FHE, shifting is faster than multiplying / dividing
+            return self
+                .unchecked_scalar_right_shift_parallelized(numerator, divisor.ilog2() as u64);
+        }
+
+        let mut chosen_multiplier =
+            ApproximatedMultiplier::choose(divisor, numerator_bits, numerator_bits);
+
+        let mut shift_pre = 0;
+        if chosen_multiplier.multiplier >= (1u128 << numerator_bits) && is_even(divisor) {
```

Comment:
this means scalar div is limited to < FheUint128 ?

Reply:
Scalar if limited to scalar being at most a u64.

To support u128 we need a maybe 1-2 things in U256
To support U256 we need a U512

Reply:
are we checking that condition ? I'm not sure I saw it

Reply:
There is 
```rust
        assert!(
            divisor_bits <= numerator_bits,
            "The divisor has more bits than the numerator\
                (numerator: {numerator_bits} bits, divisor: {divisor_bits} bits)"
        );
```

Which means at runtime we don a check on that the scalar  type is at not bigger than the number of bits in the ciphertext

And The bound on the scalar is `u64: From<T>`, and From on native type is only implemented for types where the convertion is guaranteed to be lossless (so u8->u64, u16 -> u64 but not u128 -> u64)

Reply:
I think u64 is plenty, could be good to have a backlog issue in the style "consider supporting scalar div by u128"